chromadb/test/test_client.py

```diff
@@ -70,3 +78,196 @@ def test_http_client_with_inconsistent_port_settings() -> None:
+    )
+    retry_session.heartbeat()
+
+
+def test_retry_on_503_exceeding_max(
+    httpserver: HTTPServer, retry_session: chromadb.api.ClientAPI
+) -> None:
+    httpserver.expect_request("/api/v1").respond_with_data(
+        "Service Unavailable", status=503
+    )
+
+    with pytest.raises(Exception, match="Max retries exceeded with url"):
+        retry_session.heartbeat()
+
+
+def test_no_retry_on_400(
+    httpserver: HTTPServer, retry_session: chromadb.api.ClientAPI
+) -> None:
+    httpserver.expect_request("/api/v1").respond_with_data("Bad Request", status=400)
+
+    with pytest.raises(Exception, match="Bad Request"):
+        retry_session.heartbeat()
+
+
+def test_no_retry_on_400_with_custom_retry(
+    httpserver: HTTPServer, retry_session_with_custom_retry: chromadb.api.ClientAPI
+) -> None:
+    httpserver.expect_request("/api/v1").respond_with_data("Bad Request", status=400)
+
+    with pytest.raises(Exception, match="Bad Request"):
+        retry_session_with_custom_retry.heartbeat()
+
+
+def test_no_retry_on_429_with_custom_retry(
+    httpserver: HTTPServer, retry_session_with_custom_retry: chromadb.api.ClientAPI
+) -> None:
+    httpserver.expect_request("/api/v1").respond_with_data(
+        "Too Many Requests", status=429
+    )
+
+    with pytest.raises(Exception, match="Too Many Requests"):
+        retry_session_with_custom_retry.heartbeat()
+
+
+def test_no_retry_on_504_with_custom_retry(
+    httpserver: HTTPServer, retry_session_with_custom_retry: chromadb.api.ClientAPI
+) -> None:
+    httpserver.expect_oneshot_request("/api/v1").respond_with_data(
+        "Gateway Timeout", status=504
+    )
+    httpserver.expect_oneshot_request("/api/v1").respond_with_data(
+        "Gateway Timeout", status=504
+    )
+    httpserver.expect_oneshot_request("/api/v1").respond_with_data(
+        json.dumps({"nanosecond heartbeat": 1715365335819568533}), status=200
+    )
+    retry_session_with_custom_retry.heartbeat()
+
+
+def server(max_retries: int = 1) -> None:
+    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+    s.bind(("localhost", 9999))
+    s.listen(1)
+    retries = 0
+    while retries < max_retries:
+        conn, _ = s.accept()
+        conn.close()  # Close the connection immediately
+        retries += 1
+
+
+@pytest.fixture
+def connect_retries() -> int:
+    return 3
+
+
+@pytest.fixture
+def local_tcp_server(connect_retries: int) -> Generator[None, None, None]:
+    import threading
+
+    t = threading.Thread(target=server, args=(connect_retries,))
+    t.start()
+    yield
+    t.join()
+
+
+def test_with_connection_error(
+    local_tcp_server: Optional[Any], connect_retries: int
+) -> None:
+    start_time = time.time()
+    with pytest.raises(Exception):
+        chromadb.HttpClient(
+            host="localhost",
+            port=9999,
+            retry=RetryStrategy(connect=connect_retries - 1, backoff_factor=1),
+        )
+        assert (
+            time.time() - start_time > 4
```

Comment:
sorry, not super clear to me, does this work because `.heartbeat()` is called when initing the client?